web/book/src/internals/modules.md

```diff
@@ -0,0 +1,324 @@
+    # _my_project.prql
+
+    module sales {
+        # sales.prql
+    }
+
+    module projections {
+        # projections/_projections.prql
+
+        module year_2023 {
+            # projections/year_2023.prql
+        }
+
+        module year_2024 {
+            # projections/year_2024.prql
+        }
+    }
+}
+```
+
+If module `my_project.sales` wants to add a submodule `util`, it has to be
+converted to a directory modules. This means that it has to be moved to
+`sales/_sales.prql`. The new module would reside in `sales/util.prql`.
+
+The annotated layout is not the only possible layout for this module structure,
+since any of the modules `sales`, `year_2023` or `year_2024` could be converted
+into a directory module with zero sub-modules.
+
+Point 4 of design goals means that each declaration within a project has a
+single fully-qualified name within this project. This is ensured by strict rules
+regarding importing files and the fact that the module structure is a tree.
+
+## Declaration order
+
+The order of declarations in a module holds no semantic value, except the "last
+`main` variable".
+
+References between modules can be cyclic.
+
+```
+module mod_a {
+    let decl_a_1 = ...
+    let decl_a_2 = (from mod_b.decl_b | take 10)
+}
+module mod_b {
+    let decl_b = (from mod_a.decl_a | take 10)
+}
+```
+
+References between variable declarations cannot be cyclic.
+
+```
+let decl_a = (from decl_b)
+let decl_b = (from decl_a) # error: cyclic reference
+```
+
+```
+module mod_a {
+    let decl_a = (from mod_b.decl_b)
+}
+module mod_b {
+    let decl_b = (from mod_a.decl_a) # error: cyclic reference
+}
+```
+
+## Compiler interface
+
+`prql-compiler` provides two interfaces for compiling files.
+
+**Multi-file interface** requires three arguments:
+
+- path to the file containing the module which is the root of the module
+  structure,
+- identifier of the pipeline that should be compiled to RQ (this can also be an
+  identifier of a module that has a `main` pipeline) and,
+- a "file loader", which can load files on-demand.
+
+The path to the root module can be automatically detected by searching for
+`.prql` files starting with `_` in the current working directory.
+
+Example prqlc usage:
+
+```
+$ prqlc compile _project.prql sales.projections.orders_2024
+$ prqlc compile sales.projections.orders_2024
+```
+
+**Single-file interface** requires a single argument; the PRQL source. Any
+attempts to load modules in this mode result in compilation errors. This
+interface is needed, for example, when integrating the compiler with a database
+connector (i.e. JDBC) where no other files can be loaded.
+
+## Built-in module structure
+
+> Work In Progress
+
+```
+# root module of every project
+module project {
+	module std {
+		func sum a -> ...
+		func mean a -> ...
+	}
+
+	module default_db {
+		# all inferred tables and defined CTEs
+	}
+
+	let main = (
+		from t = tracks
+		select [track_id, title]
+	)
+}
+```
+
+## Example
```

Comment:
I agree with that it should just directly load a file with the same name instead of look inside a subdirectory.

I am a bit skeptical against having imports from parent directories though (as you mention with `../`). I lean towards that files should only be able to use other files in the same directory and child directories but not in parent directories.

Reply:
> I lean towards that files should only be able to use other files in the same directory and child directories but not in parent directories.

Surely this would be super restrictive — any sort of utils package wouldn't be possible, unless every file was in the same path? Are there any other languages that do this? (Or am I misunderstanding the point?)

Reply:
> Why place the module foo declarations in the directory modules, rather than have import statements near the code that uses them?

Because with this proposal:
- you only "import"/"declare" a module once and then it's in the module structure where you can use it simply by referencing that name, no `use` or `module` statements required,
- you don't need a `import ../foo` (i.e. import parent) statement, because modules are imported/declared from root down (but you can still reference parent modules after they are imported/declared),
- refactoring is simpler, since there is a single import/declaration of a module,

> How do we discriminate between from foo.bar where foo is a DB schema vs. a module (assuming I eventually solve 
Resolve opaqueness of idents #1535)? If we have use foo in the file itself, it's explicit.

Let's discuss that after this RFC is done. I have an idea how to do it on top of modules and declarations.

Reply:
> Surely this would be super restrictive — any sort of utils package wouldn't be possible, unless every file was in the same path? Are there any other languages that do this? (Or am I misunderstanding the point?)

C# and Java imports using namespaces so nothing can be from a parent directory. PHP lets you [`include`](https://www.php.net/manual/en/function.include.php) from any file path including relative paths which may be parent directories. JavaScript (Node.js) lets you import from parent directories. I am not sure about Python, but I don't think it is common. I also worry about the security impact of importing from parent directories.

In the PHP world when using Composer it is common that imports from your own code is in your `your_project/src/` directory, and third-party code from a `your_project/vendor/` directory.

Reply:
> Because with this proposal:
> 
>     * you only "import"/"declare" a module once and then it's in the module structure where you can use it simply by referencing that name, no `use` or `module` statements required,
> 
>     * you don't need a `import ../foo` (i.e. import parent) statement, because modules are imported/declared from root down (but you can still reference parent modules after they are imported/declared),

The main issue I see is that the action is quite far from the consequence. When we're missing an item in python, it's fairly obvious where the item should be:

```
from ..utils import read_file
```

"Couldn't find read_file in utils" — OK great

But in this design, what do we say if something isn't found? There are so many places we might be missing a declaration. The user needs to understand the module system well enough to then go and check each parent in the hierarchy for the missing item.

>     * refactoring is simpler, since there is a single import/declaration of a module,

Indeed — I can imagine refactorings being quite complicated, because the compiler can no longer tell you "we can't find this item here". I'm fine pushing the conversation re "DB schema vs. a module", but this will further confuse things, since `from foo.bar` will also be legitimate SQL even when `foo` isn't available as a module.

Or are my fears overblown? I can definitely see that it's an elegant system when everything is ideal — but I can also imagine it being quite confusing with newer users or messier projects.

---

My secondary — much more minor — concern is that the `module` declarations aren't doing _that_ much. IIUC they can only refer to a single file location. Is requiring them (relative to just finding them in the file tree) going to make things clearer?

---

To put this into the example, I would do something more like python, with a touch of rust, and do:

```
# _project.prql
```

```
# employees.prql
let employees = (...)
let salaries = (...)
let departments = (...)
```

```
# sales/_sales.prql
let revenue_by_source = (...)
```

```
# sales/orders.prql
use ..employees

let current_year = (...)
let archived = (...)
let by_employee = (from orders | join employees.employees ...)
```

```
# sales/projections.prql
use ..orders

let orders_2023 = (from orders.current_year | append orders.archived)
let orders_2024 = (...)
```

```
# util.prql
func pretty_print_num col -> (...)
```

Reply:
That's a fair point, declarations are far from where they are then used. In essence, the problem with my proposal is that "the module hierarchy" is a whole concept you have to know about as opposed to just knowing how to pull a declaration from one file into another.

I've purposefully made the module system repr in filesystem normalized, so we can have very good hints about where to put your module declarations:

```
select [sales.utils.my_func my_column]
              -----
                 \___ name utils does not exist
Hint: add `module utils` to `sales/_sales.prql`
```

```
module orders
       ------
          \__ cannot load sales/orders.prql or sales/orders/_orders.prql
```

My point is that knowing about "the module hierarchy" is not a requirement for anyone, because we can provide hints to guide users until it makes sense to them.

----

My fear is that it will end up having way-too much `use`s with the extreme case being:

```
# _project.prql
```

```
# employees.prql
use ..sales
use ..util

let employees = (...)
let salaries = (...)
let departments = (...)
```

```
use projections
use orders
use ..employees
use ..utils

# sales/_sales.prql
let revenue_by_source = (...)
```

```
# sales/orders.prql
use ..
use ..projections
use ...employees
use ...utils


let current_year = (...)
let archived = (...)
let by_employee = (from orders | join employees.employees ...)
```

```
# sales/projections.prql
use ..
use ..orders
use ...employees
use ...utils

let orders_2023 = (from orders.current_year | append orders.archived)
let orders_2024 = (...)
```

```
# util.prql
func pretty_print_num col -> (...)
```

Reply:
> I've purposefully made the module system repr in filesystem normalized, so we can have very good hints about where to put your module declarations:
> 
> ```
> select [sales.utils.my_func my_column]
>               -----
>                  \___ name utils does not exist
> Hint: add `module utils` to `sales/_sales.prql`
> ```

I agree this is a very good hint. But given how we would search upwards for a module, if `sales` isn't found — i.e. a "root" item, there could be lots of places that it could be.

One option would be to do completely automatic imports — no need for `module`, no need for `use`. I might even prefer that to only having `module` — which seems to ask for more than gives, at least until we had non-file `module`s.

My synthesis after thinking about this is:
- Just using a file structure is simpler than offering non-file modules (we agree)
- `use` statements are explicit, reduce confusion, not _that_ costly
- Given the language goals & users, we should favor simplicity. And imports are often pits of confusion.
- I do think your example of the hint is quite good, but I'm still worried about the ambiguity because of the hierarchy search, and whether the hint is so good that the language should infer it itself

WDYT?

Any thoughts from others? @eitsupi & @snth ? (Note — please read the comparative examples carefully, don't make the same mistake I did...)

Reply:
Nobody wants to weigh-in here :(

> I do think your example of the hint is quite good, but I'm still worried about the ambiguity because of the hierarchy search, and whether the hint is so good that the language should infer it itself

Good point, it may be more confusing than helpful.

> One option would be to do completely automatic imports — no need for module, no need for use. I might even prefer that to only having module — which seems to ask for more than gives, at least until we had non-file modules.

That's probably the safest way to go here. It's also the easiest to implement, so would't be wasting much work if we decide to change later.

Reply:
> Nobody wants to weigh-in here :(

Yeah, one issue with these big discussions is that it's more difficult for folks to weigh in on some smaller but important issues — there are 114 comments — and it's difficult to know how much context is required for weighing in. I don't know a great way around it...

> That's probably the safest way to go here. It's also the easiest to implement, so would't be wasting much work if we decide to change later.

Great, very much agree. 

I also think that approach is the best way of getting more feedback — merge this with a path forward, and once we start building, folks will raise anything sufficient important...